B3 should create rotates from the canonical sequence of BitOr + shifts

ddegazio opened the pull request:
#### 7ed45519e747e16527bc774b25fe49feb8ca77a1
<pre>
B3 should create rotates from the canonical sequence of BitOr + shifts
<a href="https://bugs.webkit.org/show_bug.cgi?id=196366">https://bugs.webkit.org/show_bug.cgi?id=196366</a>
<a href="https://rdar.apple.com/140373792">rdar://140373792</a>

Reviewed by NOBODY (OOPS!).

Adds recognition of the x &lt;&lt; n | x &gt;&gt;&gt; (32|64)-n idiom (and all its variants) to the B3
strength reduction phase, lowering to a bitwise rotate left or rotate right. Bitwise
rotates are common in cryptography, and this pattern appears in hot loops in the
crypto-sha1-SP and crypto-md5-SP benchmarks of JetStream 2, so it&apos;s probably worthwhile
to lower this more efficiently.

* JSTests/wasm/stress/bitwise-rotate-idioms.js:
* Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
* Source/JavaScriptCore/b3/testb3_1.cpp:
* Source/JavaScriptCore/b3/testb3_1.h:
* Source/JavaScriptCore/b3/testb3_8.cpp:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7ed45519e747e16527bc774b25fe49feb8ca77a1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77588 "Build is in progress. Recent messages:OS: Ventura (13.6.7), Xcode: 14.3") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56624 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30503 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/82190 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28869 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65772 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4920 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60783 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18770 "Passed tests") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80654 "Build is in progress. Recent messages:OS: Ventura (13.6.7), Xcode: 14.3") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50758 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66590 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41055 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48162 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/24085 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27194 "Built successfully") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/70780 "Found 1 new JSC stress test failure: stress/get-by-val-hoist-above-structure.js.ftl-no-cjit-validate-sampling-profiler (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69269 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24423 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83578 "Build is in progress. Recent messages:Running configuration; Running apply-patch; Checked out pull request; Compiled WebKit") | 
| [❌ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/76870 "Found 2 new JSC stress test failures: stress/attribute-custom-accessor.js.dfg-eager-no-cjit-validate, wasm.yaml/wasm/modules/wasm-imports-js-exports.js.wasm-no-cjit (failure)") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4968 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3387 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69023 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5130 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66556 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68308 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12298 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10402 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/99118 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4915 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21666 "Found 7 new JSC stress test failures: wasm.yaml/wasm/stress/bitwise-rotate-idioms.js.default-wasm, wasm.yaml/wasm/stress/bitwise-rotate-idioms.js.wasm-bbq, wasm.yaml/wasm/stress/bitwise-rotate-idioms.js.wasm-collect-continuously, wasm.yaml/wasm/stress/bitwise-rotate-idioms.js.wasm-eager, wasm.yaml/wasm/stress/bitwise-rotate-idioms.js.wasm-eager-jettison, wasm.yaml/wasm/stress/bitwise-rotate-idioms.js.wasm-no-cjit, wasm.yaml/wasm/stress/bitwise-rotate-idioms.js.wasm-slow-memory (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4934 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8372 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6693 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
```diff
@@ -3405,6 +3411,110 @@ class ReduceStrength {
+    // Into this: RotR(x1, x2)
+    // This pattern is widely used to represent bitwise rotation in languages that don't support it natively,
+    // and appears in hot loops in cryptography algorithms, so we should try to match it when possible.
+    bool handleBitRotation()
```
nit: Maybe we can add `ASSERT(m_value->opcode() == BitOr || m_value->opcode() == BitXor);` at the beginning.

ddegazio replied:
I'll add this in the next push.

hyjorc1 commented on Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
```diff
@@ -3405,6 +3411,110 @@ class ReduceStrength {
+            if (m_value->type() == Int32
+                && shl->child(1)->hasInt()
+                && shr->child(1)->hasInt()
+                && shl->child(1)->asInt() + shr->child(1)->asInt() == 32
```
We should check both are positive since `-1 + 33 == 32`. And check overflow as well.

ddegazio replied:
B3 assumes the shift semantics are such that the shift amount is modulo the number of bits, so in the case we have `-1` and `33`, then `-1 mod 32 = 31` and `33 mod 32 = 1` and `31` and `1` are a valid pair to apply this optimization to. So I think this is fine.

We should definitely check if the sum overflows though, good catch.

hyjorc1 commented on Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
```diff
@@ -3405,6 +3411,110 @@ class ReduceStrength {
+            if (m_value->type() == Int64
+                && shl->child(1)->hasInt()
+                && shr->child(1)->hasInt()
+                && shl->child(1)->asInt() + shr->child(1)->asInt() == 64
```
Ditto.

hyjorc1 commented on Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
```diff
@@ -3405,6 +3411,110 @@ class ReduceStrength {
+            if (isRotR != isRotL
+                && ((m_value->type() == Int64 && sub->child(0)->isInt(64))
+                    || (m_value->type() == Int32 && sub->child(0)->isInt(32)))) {
+                Value* x = shl->child(0);
```
nit: Maybe we can use `x1` for `x` for readability since it aligns to the comment shown above.

ddegazio replied:
Ah, sure, that comment is newer than this code. I'll change it in the next push.

hyjorc1 commented on Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
```diff
@@ -3405,6 +3411,110 @@ class ReduceStrength {
+                // these expressions are being used as the shift amount, all of this
+                // arithmetic is implicitly mod 32 or mod 64 - so we can safely
+                // ignore intermediate Trunc nodes.
+                Value* n = sub->child(1);
```
This `n` is `x2` right? Then, it must be a constant since we have to check the overflow of `sub`.

hyjorc1 replied:
We also need to check whether constant folding is applied or not in this phase. If it is, then the `sub` case should not be observable here.

ddegazio replied:
I'll change this name too, but in this case it's specifically non-constant - or, well, I suppose it could be, but we would indeed expect the `Sub` to fold away if it was. I think since B3 guarantees wrapping semantics for `Sub`, it's okay if the difference overflows - since overflow here is basically wrapping the shift amount by a large power-of-two modulus, and then the shift itself wraps the shift amount to a smaller power-of-two modulus, I think the overall result can't be observably different.

ddegazio committed:
B3 should create rotates from the canonical sequence of BitOr + shifts
https://bugs.webkit.org/show_bug.cgi?id=196366
rdar://140373792

Reviewed by NOBODY (OOPS!).

Adds recognition of the x << n | x >>> (32|64)-n idiom (and all its variants) to the B3
strength reduction phase, lowering to a bitwise rotate left or rotate right. Bitwise
rotates are common in cryptography, and this pattern appears in hot loops in the
crypto-sha1-SP and crypto-md5-SP benchmarks of JetStream 2, so it's probably worthwhile
to lower this more efficiently.

* JSTests/wasm/stress/bitwise-rotate-idioms.js:
* Source/JavaScriptCore/b3/B3ReduceStrength.cpp:
* Source/JavaScriptCore/b3/testb3_1.cpp:
* Source/JavaScriptCore/b3/testb3_1.h:
* Source/JavaScriptCore/b3/testb3_8.cpp:

justinmichaud reviewed (approved):
r=me
